### PR TITLE
Backing store data class: Default to schema hash when `name` is null.

### PR DESCRIPTION
**Pull request by alxrsngrtn**

### CODE REVIEW

File: java/arcs/core/storage/StorageKeyFactory.kt [hunk 1]
```diff
@@ -52,8 +52,9 @@ abstract class StorageKeyFactory(val protocol: String, val capabilities: Capabil
         override val arcId: ArcId,
         override val entitySchema: Schema
     ) : StorageKeyOptions {
-        override val unique: String =
-        requireNotNull(entitySchema).name?.name ?: entitySchema.hash
+        override val unique: String = with(entitySchema.name?.name) {
```
Comment (mmandlis) [inline java/arcs/core/storage/StorageKeyFactory.kt:55]: when will `entitySchema.name` be null? 
also, could you, please, add tests to cover all these edge cases? 
thanks.

Comment (piotrswigon) [reply java/arcs/core/storage/StorageKeyFactory.kt:55]: Also, can we understand what is happening first before making changes to code?

Do we know why it's null? Do we know if this fixes Satakshis problem?

Comment (alxrsngrtn) [reply java/arcs/core/storage/StorageKeyFactory.kt:55]: I have a few suspicions as to why this fix should work and what is going on. I'm currently testing them out in cider. Currently, it seems like this will fix the issue -- but is it the right fix? I'll keep digging.  

@mmandlis  `entitySchema.name` can be null anytime we have a schema without any names. For instance, 
```
particle TimestampMsToDayRedactor in '.TimestampMsToDayRedactor'
  input: reads [~a with {timestampInMs: Number}]
```
Here, the schema.name should be null. 

Test cases added in 2860f0f.

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+/** Tests for [StorageKeyFactory]. */
+@RunWith(JUnit4::class)
+class StorageKeyFactoryTest {
+
+    @Test
+    fun backingStore_namelessEntity() {
+        val emptySchema = Schema.EMPTY
```
Comment (mmandlis) [inline javatests/arcs/core/storage/StorageKeyFactoryTest.kt:18]: nit Schema.EMPTY is initialized with `setOf()` for names, but maybe worth asserting on that here, or adding an explicit schema with empty set of names (for the sake of the case EMPTY changes).

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        val storeOptionsEmpty = StorageKeyFactory.BackingStorageKeyOptions(
+            arcId, emptySchema
+        )
+
+        val storeOptionsEmptyName = StorageKeyFactory.BackingStorageKeyOptions(
+            arcId, emptyNamedSchema
+        )
+
+        assertThat(storeOptionsEmpty.unique).isEqualTo(emptySchema.hash)
+        assertThat(storeOptionsEmptyName.unique).isEqualTo(emptyNamedSchema.hash)
```
Comment (mmandlis) [inline javatests/arcs/core/storage/StorageKeyFactoryTest.kt:34]: can you, please, add one more test case with a non-empty name as well?

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt (hunk 2 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/storage/StorageKeyFactoryTest.kt:18]: 👍  will decouple EMPTY from this test.

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt (hunk 3 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/storage/StorageKeyFactoryTest.kt:34]: Added another test method.

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt [hunk 4]
```diff
@@ -13,9 +13,27 @@ import org.junit.runners.JUnit4
...
+            arcId, namedSchema
+        )
+
+        assertThat(storeOptionsEmptyName.unique).isEqualTo(namedSchema.name?.name)
+    }
+
     @Test
     fun backingStore_namelessEntity() {
-        val emptySchema = Schema.EMPTY
+        val emptySchema = Schema(emptySet(),
```
Comment (mmandlis) [inline javatests/arcs/core/storage/StorageKeyFactoryTest.kt:33]: nit - i think `emptySet(),` needs to go on a new line as well in kt (here and below)

File: javatests/arcs/core/storage/StorageKeyFactoryTest.kt [hunk 5]
```diff
@@ -13,9 +13,27 @@ import org.junit.runners.JUnit4
...
+            SchemaFields(emptyMap(), emptyMap()),
+            "SomeHash"
+        )
+        val arcId = ArcId.newForTest("myId")
+
+        val storeOptionsEmptyName = StorageKeyFactory.BackingStorageKeyOptions(
+            arcId, namedSchema
+        )
+
+        assertThat(storeOptionsEmptyName.unique).isEqualTo(namedSchema.name?.name)
```
Comment (mmandlis) [inline javatests/arcs/core/storage/StorageKeyFactoryTest.kt:28]: maybe instead use `requireNotNull(namedSchema.name).name`, or even just "Foo" here, to avoid a double bug when it's empty in both :)
